flang/test/Lower/forall.f90

```diff
@@ -1165,3 +1165,101 @@ end function f
+  ! CHECK: %[[VAL_9:.*]] = fir.coordinate_of %[[VAL_1]], %[[VAL_8]] : (!fir.ref<!fir.type<_QFforall_with_allocatable2Tt{i:i32,arr:!fir.box<!fir.heap<!fir.array<?xf32>>>}>>, !fir.field) -> !fir.ref<!fir.box<!fir.heap<!fir.array<?xf32>>>>
+  ! CHECK: %[[VAL_10:.*]] = fir.load %[[VAL_9]] : !fir.ref<!fir.box<!fir.heap<!fir.array<?xf32>>>>
+  ! CHECK: %[[VAL_11:.*]] = constant 0 : index
+  ! CHECK: %[[VAL_12:.*]]:3 = fir.box_dims %[[VAL_10]], %[[VAL_11]] : (!fir.box<!fir.heap<!fir.array<?xf32>>>, index) -> (index, index, index)
+  ! CHECK: %[[VAL_13:.*]] = fir.box_addr %[[VAL_10]] : (!fir.box<!fir.heap<!fir.array<?xf32>>>) -> !fir.heap<!fir.array<?xf32>>
+  ! CHECK: %[[VAL_14:.*]] = fir.shape_shift %[[VAL_12]]#0, %[[VAL_12]]#1 : (index, index) -> !fir.shapeshift<1>
+  ! CHECK: %[[VAL_15:.*]] = constant 1 : index
+  ! CHECK: %[[VAL_16:.*]] = fir.slice %[[VAL_12]]#0, %[[VAL_12]]#1, %[[VAL_15]] : (index, index, index) -> !fir.slice<1>
+  ! CHECK: %[[VAL_17:.*]] = fir.array_load %[[VAL_13]](%[[VAL_14]]) {{\[}}%[[VAL_16]]] : (!fir.heap<!fir.array<?xf32>>, !fir.shapeshift<1>, !fir.slice<1>) -> !fir.array<?xf32>
+  ! CHECK: %[[VAL_18:.*]] = fir.array_load %[[arg0]] : (!fir.box<!fir.array<?xf32>>) -> !fir.array<?xf32>
+  ! CHECK: %[[VAL_20:.*]] = constant 5 : i32
+  ! CHECK: %[[VAL_21:.*]] = fir.convert %[[VAL_20]] : (i32) -> index
+  ! CHECK: %[[VAL_22:.*]] = constant 15 : i32
+  ! CHECK: %[[VAL_23:.*]] = fir.convert %[[VAL_22]] : (i32) -> index
+  ! CHECK: %[[VAL_24:.*]] = constant 1 : index
+  real :: a1(:)
+  type t
+     integer :: i
+     real, allocatable :: arr(:)
+  end type t
+  type(t) :: thing
+
+  ! CHECK: %[[V_0:.*]] = fir.do_loop %[[V_1:.*]] = %[[VAL_21]] to %[[VAL_23]] step %[[VAL_24]] unordered iter_args(%[[V_2:.*]] = %[[VAL_17]]) -> (!fir.array<?xf32>) {
+  ! CHECK: %[[V_3:.*]] = fir.convert %[[V_1]] : (index) -> i32
+  ! CHECK: fir.store %[[V_3]] to %[[VAL_0]] : !fir.ref<i32>
+  ! CHECK: %[[V_4:.*]] = fir.load %[[VAL_0]] : !fir.ref<i32>
+  ! CHECK: %[[V_5:.*]] = fir.convert %[[V_4]] : (i32) -> i64
+  ! CHECK: %[[V_6:.*]] = fir.convert %[[V_5]] : (i64) -> index
+  ! CHECK: %[[V_7:.*]] = fir.array_fetch %[[VAL_18]], %[[V_6]] : (!fir.array<?xf32>, index) -> f32
+  ! CHECK: %[[V_8:.*]] = fir.load %[[VAL_0]] : !fir.ref<i32>
+  ! CHECK: %[[V_9:.*]] = fir.convert %[[V_8]] : (i32) -> i64
+  ! CHECK: %[[V_10:.*]] = fir.convert %[[V_9]] : (i64) -> index
+  ! CHECK: %[[V_11:.*]] = fir.array_update %[[V_2]], %[[V_7]], %[[V_10]] : (!fir.array<?xf32>, f32, index) -> !fir.array<?xf32>
+  ! CHECK: fir.result %[[V_11]] : !fir.array<?xf32>
+  ! CHECK: }
+  forall (i=5:15)
+     thing%arr(i) = a1(i)
+  end forall
+  ! CHECK: fir.array_merge_store %[[VAL_17]], %[[V_0]] to %[[VAL_13]]{{\[}}%[[VAL_16]]] : !fir.array<?xf32>, !fir.array<?xf32>, !fir.heap<!fir.array<?xf32>>, !fir.slice<1>
+  ! CHECK: return
+end subroutine forall_with_allocatable2
```

Comment:
This test does not seem to exerce the allcoatable reassignment semantics (assignment to whole allocatable). Looking at the interface change of  `createAllocatableArrayAssignment` and the title of the PR, I though this was now working.

Is the following program working with this patch ? :
```
subroutine forall_with_allocatable3(a1)
  real :: a1(:)
  type t
     real, allocatable :: arr(:)
  end type t
  type(t) :: thing(20)
  forall (i=5:15)
     thing(i)%arr = a1
  end forall
end
```

If yes, it would be worth adding it to the tests.

BTW, you may want to split the test file at some point into forall1.f90 forall2.f90, It is becoming quite large.

Reply:
Your example will require refactoring genExprMutableBox, which wasn't a part of this PR.

Reply:
OK, got it thanks.